// @nuxt-themes/tokens/dist/module.ts.mjs
import { defineNuxtModule, createResolver, installModule } from '@nuxt/kit';

const envModules = {
  colorMode: process?.env?.THEME_DEV_COLOR_MODE_PATH || "@nuxtjs/color-mode",
  pinceau: process?.env?.THEME_DEV_PINCEAU_PATH || "pinceau/nuxt"
};
const module = defineNuxtModule({
  meta: {
    name: "@nuxt-themes/tokens",
    configKey: "tokens"
  },
  async setup(_, nuxt) {
    const modulePath = createResolver(import.meta.url);
    nuxt.hook(
      "pinceau:options",
      (options) => {
        options.configLayers = options?.configLayers || [];
        options.configLayers.push({
          cwd: modulePath.resolve(),
          configFileName: "tokens.config"
        });
      }
    );
    nuxt.hook("prepare:types", (opts) => {
      const tsConfig = opts.tsConfig;
      tsConfig.compilerOptions = tsConfig.compilerOptions || {};
      tsConfig.compilerOptions.paths = tsConfig.compilerOptions.paths || {};
      tsConfig.compilerOptions.paths["@nuxt-themes/tokens/tokens"] = [modulePath.resolve("./tokens.config.ts")];
      tsConfig.compilerOptions.paths["@nuxt-themes/tokens/config"] = [modulePath.resolve("./tokens.config.ts")];
    });
    await installModule(envModules.colorMode, { classSuffix: "" });
    await installModule(envModules.pinceau, { configFileName: "tokens.config", colorSchemeMode: "class" });
  }
});

export { module as default };
