// @nuxt-themes/elements/tokens.config.ts
import { defineTheme } from 'pinceau'
import theme from '@nuxt-themes/tokens/tokens'

export default defineTheme({
  color: {
    primary: theme.color.teal
  },

  container: {
    maxWidth: '80rem',
    padding: {
      mobile: '{space.4}',
      xs: '{space.4}',
      sm: '{space.6}',
      md: '{space.6}'
    }
  },

  backdrop: {
    filter: 'saturate(200%) blur(20px)',
    background: {
      // TODO: rgba 80% from color.white
      initial: '#fffc',
      // TODO: rgba 80% from color.black
      dark: '#0c0d0ccc'
    }
  },

  text: {
    color: {
      primary: {
        initial: '{color.gray.900}',
        dark: '{color.gray.50}'
      },
      secondary: {
        initial: '{color.gray.500}',
        dark: '{color.gray.400}'
      },
      secondaryHover: {
        initial: '{color.gray.700}',
        dark: '{color.gray.200}'
      }
    }
  },

  borders: {
    primary: {
      default: {
        initial: '{color.gray.100}',
        dark: '{color.gray.900}'
      },
      hover: {
        initial: '{color.gray.200}',
        dark: '{color.gray.800}'
      }
    },
    secondary: {
      default: {
        initial: '{color.gray.200}',
        dark: '{color.gray.800}'
      },
      hover: {
        initial: '',
        dark: ''
      }
    },
  },

  surface: {
    background: {
      base: {
        initial: '{color.gray.100}',
        dark: '{color.gray.900}'
      }
    }
  },

  states: {
    primary: {
      color: {
        primary: {
          initial: '{color.primary.600}',
          dark: '{color.primary.400}'
        },
        secondary: {
          initial: '{color.primary.700}',
          dark: '{color.primary.200}'
        }
      },
      backgroundColor: {
        primary: {
          initial: '{color.primary.50}',
          dark: '{color.primary.900}'
        },
        secondary: {
          initial: '{color.primary.100}',
          dark: '{color.primary.800}'
        }
      },
      borderColor: {
        primary: {
          initial: '{color.primary.100}',
          dark: '{color.primary.800}'
        },
        secondary: {}
      }
    },
    info: {
      color: {
        primary: {
          initial: '{color.blue.500}',
          dark: '{color.blue.400}'
        },
        secondary: {
          initial: '{color.blue.600}',
          dark: '{color.blue.200}'
        }
      },
      backgroundColor: {
        primary: {
          initial: '{color.blue.50}',
          dark: '{color.blue.900}'
        },
        secondary: {
          initial: '{color.blue.100}',
          dark: '{color.blue.800}'
        }
      },
      borderColor: {
        primary: {
          initial: '{color.blue.100}',
          dark: '{color.blue.800}'
        },
        secondary: {}
      }
    },
    success: {
      color: {
        primary: {
          initial: '{color.green.500}',
          dark: '{color.green.400}'
        },
        secondary: {
          initial: '{color.green.600}',
          dark: '{color.green.200}'
        }
      },
      backgroundColor: {
        primary: {
          initial: '{color.green.50}',
          dark: '{color.green.900}'
        },
        secondary: {
          initial: '{color.green.100}',
          dark: '{color.green.800}'
        }
      },
      borderColor: {
        primary: {
          initial: '{color.green.100}',
          dark: '{color.green.800}'
        },
        secondary: {}
      }
    },
    warning: {
      color: {
        primary: {
          initial: '{color.yellow.600}',
          dark: '{color.yellow.400}'
        },
        secondary: {
          initial: '{color.yellow.700}',
          dark: '{color.yellow.200}'
        }
      },
      backgroundColor: {
        primary: {
          initial: '{color.yellow.50}',
          dark: '{color.yellow.900}'
        },
        secondary: {
          initial: '{color.yellow.100}',
          dark: '{color.yellow.800}'
        }
      },
      borderColor: {
        primary: {
          initial: '{color.yellow.100}',
          dark: '{color.yellow.800}'
        },
        secondary: {}
      }
    },
    danger: {
      color: {
        primary: {
          initial: '{color.red.500}',
          dark: '{color.red.300}'
        },
        secondary: {
          initial: '{color.red.600}',
          dark: '{color.red.200}'
        }
      },
      backgroundColor: {
        primary: {
          initial: '{color.red.50}',
          dark: '{color.red.900}'
        },
        secondary: {
          initial: '{color.red.100}',
          dark: '{color.red.800}'
        }
      },
      borderColor: {
        primary: {
          initial: '{color.red.100}',
          dark: '{color.red.800}'
        },
        secondary: {}
      }
    }
  },

  utils: {
    stateColors: (value: any) => {
      return {
        color: `{states.${value}.color.primary}`,
        backgroundColor: `{states.${value}.backgroundColor.primary}`,
        borderColor: `{states.${value}.borderColor.primary}`,
        ':deep(p code)': {
          color: `{states.${value}.color.secondary}`,
          backgroundColor: `{states.${value}.backgroundColor.secondary}`
        },
        ':deep(a)': {
          borderColor: 'currentColor',
          '&:hover': {
            color: `{states.${value}.color.secondary}`,
            borderColor: 'currentColor'
          }
        }
      }
    }
  }
})
