// pinceau/dist/utils.cjs
'use strict';

const stringify = require('./shared/pinceau.7e38ce94.cjs');
const data = require('./shared/pinceau.7cd078d1.cjs');
const palette = require('./shared/pinceau.f7625e1a.cjs');
const theme = require('./shared/pinceau.c574f603.cjs');
require('vue');
require('scule');
require('chroma-js');
require('pathe');
require('postcss-nested');
require('postcss-custom-properties');
require('postcss-dark-theme-class');



exports.DARK = stringify.DARK;
exports.INITIAL = stringify.INITIAL;
exports.LIGHT = stringify.LIGHT;
exports.castValue = stringify.castValue;
exports.castValues = stringify.castValues;
exports.createTokensHelper = stringify.createTokensHelper;
exports.cssContentRegex = stringify.cssContentRegex;
exports.darkRegex = stringify.darkRegex;
exports.dtRegex = stringify.dtRegex;
exports.getResolvedSelectors = stringify.getResolvedSelectors;
exports.keyRegex = stringify.keyRegex;
exports.lightRegex = stringify.lightRegex;
exports.mqCssRegex = stringify.mqCssRegex;
exports.mqPlainRegex = stringify.mqPlainRegex;
exports.outputFileNames = stringify.outputFileNames;
exports.pathToVarName = stringify.pathToVarName;
exports.referencesRegex = stringify.referencesRegex;
exports.resolveCssProperty = stringify.resolveCssProperty;
exports.resolveCustomDirectives = stringify.resolveCustomDirectives;
exports.resolveReferences = stringify.resolveReferences;
exports.resolveVariableFromPath = stringify.resolveVariableFromPath;
exports.responsiveMediaQueryRegex = stringify.responsiveMediaQueryRegex;
exports.stringify = stringify.stringify;
exports.transformTokensToVariable = stringify.transformTokensToVariable;
exports.variantsRegex = stringify.variantsRegex;
exports.get = data.get;
exports.normalizeConfig = data.normalizeConfig;
exports.set = data.set;
exports.walkTokens = data.walkTokens;
exports.palette = palette.palette;
exports.flattenTokens = theme.flattenTokens;
exports.objectPaths = theme.objectPaths;
exports.parseVueQuery = theme.parseVueQuery;
exports.registerAliases = theme.registerAliases;
exports.registerPostCssPlugins = theme.registerPostCssPlugins;
